epochStart/bootstrap/shardStorageHandler.go

```diff
@@ -376,35 +463,39 @@ func printProcessedAndPendingMbs(processedMiniBlocks []bootstrapStorage.MiniBloc
 	}
 }
 
-func addMbToPendingList(
-	mbHandler data.MiniBlockHeaderHandler,
+func addMiniBlockToPendingList(
+	mbHeader data.MiniBlockHeaderHandler,
 	pendingMiniBlocks []bootstrapStorage.PendingMiniBlocksInfo,
 ) []bootstrapStorage.PendingMiniBlocksInfo {
 	for i := range pendingMiniBlocks {
-		if pendingMiniBlocks[i].ShardID == mbHandler.GetReceiverShardID() {
-			pendingMiniBlocks[i].MiniBlocksHashes = append(pendingMiniBlocks[i].MiniBlocksHashes, mbHandler.GetHash())
+		if pendingMiniBlocks[i].ShardID == mbHeader.GetReceiverShardID() {
+			for _, mbHash := range pendingMiniBlocks[i].MiniBlocksHashes {
+				if bytes.Equal(mbHash, mbHeader.GetHash()) {
+					return pendingMiniBlocks
```

Comment:
This method is called when a node starts from storage on this route SaveDataToStorage -> getProcessedAndPendingMiniBlocksWithScheduled -> addMiniBlocksToPending, when we want to roll back one block from the notarizedShardHeader. In this situation we need to put all these cross mini blocks which are included in this header in a pending state. Actualy we do this in two steps. First we call getProcessedAndPendingMiniBlocks and afterwards if we are in "withSchedued" state, we call updatePendingMiniBlocksForScheduled. Before the call of the second method we populate pendingsMiniBlocks through the addMiniBlocksToPending method which calls in a loop the method in charge. 

```
	for _, miniBlockHeader := range mapHashMiniBlockHeaders {
		pendingMiniBlocks = addMiniBlockToPendingList(miniBlockHeader, pendingMiniBlocks)
	}
```
We iterate all the mini blocks header from the notarizedShardheader we want to rollback, and inside the method addMiniBlockToPendingList we check if this mini block was already added as pending. If this is true we will skip the add as we don't want to have duplicated pending mini blocks. 

This is why we return the entire list which is actually one of the input parameter. In this case we don't want to modify(add an element) in the list as it is already there. 

This situation was happened/found/fixed in the system tests, because of the partial execution approach. Now a mini block header could be added in the first step (if it is partial executed) and also in the second step, when we want to roll back the notarizedHeader which contains this partial executed mb.

Reply:
Refactored a little bit here